apis/database.js

```diff
@@ -23,30 +25,60 @@ export function executeTransaction(statement, args = null) {
   });
 }
 
-export function createDatabases() {
+export async function createDatabases() {
+  await SecureStore.setItemAsync(DATABASE_VERSION_KEY, Constants.manifest.extra.localDatabaseVersion);
+
   return executeTransaction(
-    'create table if not exists locations (id integer primary key not null, key text, coordinateKey text, createdAt text, resources text, status text, uploaded int)'
+    'create table if not exists locations (id integer primary key not null, key text, coordinateKey text, createdAt text, resources text, status text, uploaded int, updated int)'
   );
 }
 
 export function clearDatabase() {
   return executeTransaction('drop table locations');
 }
 
+export function deleteLocation(key) {
+  return executeTransaction('delete from locations where key = ?', [key]);
+}
+
 export function updateUploadStatus(key, isUploaded) {
   return executeTransaction('update locations set uploaded = ? where key = ?', [isUploaded, key]);
 }
 
-export function updateLocalLocation(options) {
-  const { resources, status, key } = options;
-  // Uncomment this if user is ever able to change location position
-  // SecureStore.setItemAsync(key, JSON.parse({ longitude, latitude }));
+export function replaceLocalLocationWithRemote(remoteLocation) {
+  const { key, latitude, longitude, updated, status, resources } = remoteLocation;
 
-  return executeTransaction('update locations set resources = ?, status = ?, uploaded = 0 where key = ?', [
+  saveCoordinates(key, latitude, longitude);
+
+  return executeTransaction('update locations set status = ?, resources = ?, updated = ?, uploaded = 1 where key = ?', [
+    status,
+    JSON.stringify(resources),
+    updated,
+    key,
+  ]);
+}
+
+export async function updateLocalLocation(options, oldLocation) {
+  const { key, updated = 0 } = oldLocation;
+  const { latitude, longitude, resources, status } = options;
+
+  saveCoordinates(key, latitude, longitude);
+
+  const locationObject = createLocationObject(key, {
+    ...oldLocation,
+    ...options,
+    uploaded: false,
+    updated: oldLocation.uploaded ? oldLocation.updated + 1 : oldLocation.updated,
+  });
+
+  await executeTransaction('update locations set resources = ?, status = ?, uploaded = 0, updated = ? where key = ?', [
```

Comment:
Does this also need a transaction or to enforce that update is equal to update - 1 before it is being set?

Reply:
I'm not sure that's totally necessary. The value of `updated` will always be less than the new value. If it is less than the new value -1 then we are just making an even newer update to the local location and it should be a non-issue. Is there something else I am overlooking?

Reply:
But I already do a check to ensure the remote data hasn't been updated more recently no?

Reply:
You do, but there is still a race condition. Multiple people could try to update it at the same time which would still result in one overwriting the other if the concurrency bit is only set locally. I was expecting something to be passed up with the update to prevent races in firebase.